### PR TITLE
crumbs: init at version 0.0.3

### PR DESCRIPTION
**Pull request by Thesola10**
Let me know if there are any errors with the way this derivation is written.

###### Motivation for this change
I wanted to add this little utility to Nix to make installing it easier.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"` (None depend anyway)
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/applications/misc/crumbs/default.nix
Comment (symphorien) [inline pkgs/applications/misc/crumbs/default.nix]: please use `pname` instead of `name`.

File: pkgs/applications/misc/crumbs/default.nix
Comment (symphorien) [inline pkgs/applications/misc/crumbs/default.nix]: I think this is haskell specific, so unneeded here.

File: pkgs/applications/misc/crumbs/default.nix
Comment (symphorien) [inline pkgs/applications/misc/crumbs/default.nix]: this is already included in stdenv.

File: pkgs/applications/misc/crumbs/default.nix
Comment (symphorien) [inline pkgs/applications/misc/crumbs/default.nix]: automake and autoconf are executed when building, not at runtime, so they should go in `nativeBuildInputs` instead of `buildInputs`.
In any case, if you replace them by `autoreconfHook`, you can get rid of the custom `configurePhase`.

**Review by symphorien — CHANGES_REQUESTED** [review]
Thanks for your PR!
You forgot to include the callPackage line in `pkgs/top-level/all-package.nix` declaring crumbs. I also left a few comments to improve the derivation, but they are not so important.

**Comment by Thesola10** [comment]
I have submitted new changes applying your suggested changes. Tested the new derivation and it installs properly.

File: pkgs/applications/misc/crumbs/default.nix
Comment (Thesola10) [reply pkgs/applications/misc/crumbs/default.nix]: my commit replacing `name` with `pname` seems to have failed the package listing tests. Is there some thing I'm missing?

File: pkgs/applications/misc/crumbs/default.nix
Comment (symphorien) [inline pkgs/applications/misc/crumbs/default.nix]:
```suggestion
{ stdenv, fetchFromGitHub, autoreconfHook, gzip, ruby, bash }:
```
This is the reason of the ofborg error. You can get access to the error message by running `nix-review pr 58960 -p crumbs`:
```
error: while evaluating 'callPackageWith' at /home/symphorien/.cache/nix-review/pr-58960-1/nixpkgs/lib/customisation.nix:108:35, called from /home/symphorien/.cache/nix-review/pr-58960-1/nixpkgs/pkgs/top-level/all-packages.nix:699:12:
while evaluating 'makeOverridable' at /home/symphorien/.cache/nix-review/pr-58960-1/nixpkgs/lib/customisation.nix:67:24, called from /home/symphorien/.cache/nix-review/pr-58960-1/nixpkgs/lib/customisation.nix:112:8:
undefined variable 'autoreconfHook' at /home/symphorien/.cache/nix-review/pr-58960-1/nixpkgs/pkgs/applications/misc/crumbs/default.nix:15:7
```

Comment (Thesola10) [reply pkgs/applications/misc/crumbs/default.nix]: I didn't know about that tool, thanks

File: pkgs/applications/misc/crumbs/default.nix
Comment (tadeokondrak) [inline pkgs/applications/misc/crumbs/default.nix]:
```suggestion
      rev = "v${version}";
```
also, 0.0.3 seems to be out

**Comment by tadeokondrak** [comment]
This should ideally be rebased into one commit with the message `crumbs: init at version 0.0.2`.

File: pkgs/applications/misc/crumbs/default.nix
Comment (symphorien) [inline pkgs/applications/misc/crumbs/default.nix]: `ruby` seems unneeded by reading the homepage.

File: pkgs/applications/misc/crumbs/default.nix
Comment (symphorien) [inline pkgs/applications/misc/crumbs/default.nix]:
```suggestion
      license     = licenses.wtfpl;
```
(nitpick)

**Comment by symphorien** [comment]
I added a few more comments, but it looks good to me overall.

File: pkgs/applications/misc/crumbs/default.nix
Comment (aanderse) [inline pkgs/applications/misc/crumbs/default.nix]: Both `ruby` and `bash` can be removed as they aren't being referenced.

File: pkgs/applications/misc/crumbs/default.nix
Comment (aanderse) [inline pkgs/applications/misc/crumbs/default.nix]: Quoting not required.

**Review by aanderse — CHANGES_REQUESTED** [review]
Please rewrite into 2 commits: the first adding yourself to maintainer list, the second adding the package.

**Comment by Thesola10** [comment]
Removed a trailing comma in the includes header of `default.nix`

**Comment by aanderse** [comment]
@Thesola10 looks like you might have mixed something up on that last push:

> hash mismatch in fixed-output derivation '/nix/store/7km1gylsg0xscbp538rfq4w8rqa3zbp8-source':
> wanted: sha256:1m7fwjw74fy8pgsvjda1c7awl2y5jr0r3fb5a6v5rddym5w52s0b
> got:    sha256:0j7axygbaz3bcpjngybigk333lga67bg7g8gw47j4xx3gyc9qf2n

**Comment by Thesola10** [comment]
@aanderse there's a reason behind it. I think I had used a specific commit for reference instead of version 0.0.2, and so I pushed the wrong hash.

**Comment by aanderse** [comment]
@Thesola10 is this currently building on your system? I pulled again and got this:
> autoreconf: running: automake --add-missing --copy --force-missing
> Makefile.am: error: required file './README' not found
> autoreconf: automake failed with exit status: 1

**Comment by Thesola10** [comment]
this is very weird. noticed it also does that with `0.0.3`...

**Comment by Thesola10** [comment]
Upon second look, it seems that the `--force-missing` option that `autoreconfHook` adds is causing this. I just wrote a test `configurationPhase` locally to test that and it works again. Can I push that, or is there a cleaner way?

**Comment by Thesola10** [comment]
Also, I'm kind of weirded out by the fact that @GrahamcOfBorg doesn't seem to test whether the actual derivation can build properly...

**Comment by aanderse** [comment]
> Also, I'm kind of weirded out by the fact that @GrahamcOfBorg doesn't seem to test whether the actual derivation can build properly...

I'll be asking the bot to build your program once it builds on my machine.

> Can I push that, or is there a cleaner way?

Please rewrite the git history.

**Comment by Thesola10** [comment]
> Please rewrite the git history.

Of course. I was asking if there was a better way than to redefine `configurePhase`.

**Comment by aanderse** [comment]
> Of course. I was asking if there was a better way than to redefine `configurePhase`.

Ah... sorry I read that wrong the first time :laughing: I'll defer to @jtojnar for all `autoreconfHook` related questions though :)

**Comment by jtojnar** [comment]
You really should not override `configurePhase` it breaks things like `configureFlags`. You can call `automake` in `preConfigure`.

Or you can override the flags `autoreconfHook` sets with `autoreconfFlags`:

https://github.com/NixOS/nixpkgs/blob/05a53ecd827c5910e34171e7681c56913b29add1/pkgs/build-support/setup-hooks/autoreconf.sh#L5

**Comment by Thesola10** [comment]
@jtojnar thanks for the advice! After testing, though, what worked was setting `automake` and `autoconf` as nativeBuildInputs, and overriding `automakeFlags` to `--add-missing`.

File: pkgs/applications/misc/crumbs/default.nix
Comment (aanderse) [reply pkgs/applications/misc/crumbs/default.nix]: @Thesola10 can you please address the 0.0.3 release comment?

**Comment by jtojnar** [comment]
I would not expect `automakeFlags` to do anything, it is not defined in `nixpkgs`.

It looks like they have pregenerated `./configure` script in the repo so autotools should not be needed at all unless the script is incorrectly generated and we need to regenerate it.

**Comment by Thesola10** [comment]
> I would not expect `automakeFlags` to do anything,

You're right, it does not seem necessary. In any case, using `autoconf` and `automake` instead of `autoreconfHook` seems to work.

**Comment by Thesola10** [comment]
After another test, it would seem that configure is pre-generated. Which is weird, since I remember a previous version required to `automake` it.

**Comment by aanderse** [comment]
@Thesola10 seems like only outstanding issue is addressing the 0.0.3 comment mentioned in https://github.com/NixOS/nixpkgs/pull/58960#discussion_r272304117.

**Comment by Thesola10** [comment]
> the 0.0.3 comment
Do you mean that I need to bump the derivation to crumbs `0.0.3`? I can do that, just making sure

**Comment by Thesola10** [comment]
I think all of the discrepancies in the build process (like `autoreconfHook`) come from the difference between master and release tags in the crumbs repo.

**Comment by aanderse** [comment]
> > the 0.0.3 comment
> 
> Do you mean that I need to bump the derivation to crumbs `0.0.3`? I can do that, just making sure

I did mean that. I ran checked out your branch and it does build fine as is on 0.0.3.
While in there I happened to notice this: https://github.com/fasseg/crumbs/blob/master/crumbs-completion.bash
Any motivation to add this to your package? Seems valuable :smile:

**Comment by Thesola10** [comment]
> Any motivation to add this to your package?

Would adding it into `$out/share/bash-completion/completions` work?

**Comment by aanderse** [comment]
Here are a few examples of bash and fish shell completions searching through the codebase turned up:

https://github.com/NixOS/nixpkgs/blob/cfa3f9bb996e4c871c8574fb5302512ed4485abc/pkgs/development/tools/kubectx/default.nix
https://github.com/NixOS/nixpkgs/blob/11318d1ba2d3a06dcffa9144042c1397eff4ac9e/pkgs/applications/science/logic/elan/default.nix
https://github.com/NixOS/nixpkgs/blob/25e7f59516b2c8254df13b448a2ed5e28b848dd3/pkgs/applications/version-management/pijul/default.nix

**Comment by Thesola10** [comment]
Added the autocomplete scripts in a `postInstall` directive, since I saw these examples were overriding `installPhase` (which I kinda don't like, always prefer builtin to handwritten). The files end up where they need to be, and the derivation builds fine.
Oh and I bumped to `0.0.3` while we're at it.

**Comment by aanderse** [comment]
Looking good. The fish completions reference gfind instead of find so I made this change and then they worked:

```
  prePatch = ''
    sed -i 's|gfind|find|' crumbs-completion.fish
  '';
```
and in the postInstall
```
-    cp $src/crumbs-completion.bash $out/share/bash-completion/completions/crumbs
-    cp $src/crumbs-completion.fish $out/share/fish/vendor_completions.d/crumbs.fish
+    cp crumbs-completion.bash $out/share/bash-completion/completions/crumbs
+    cp crumbs-completion.fish $out/share/fish/vendor_completions.d/crumbs.fish
```

**Comment by Thesola10** [comment]
> and in the postInstall
> 
> ```
> -    cp $src/crumbs-completion.bash $out/share/bash-completion/completions/crumbs
> -    cp $src/crumbs-completion.fish $out/share/fish/vendor_completions.d/crumbs.fish
> +    cp crumbs-completion.bash $out/share/bash-completion/completions/crumbs
> +    cp crumbs-completion.fish $out/share/fish/vendor_completions.d/crumbs.fish
> ```

Wouldn't the reference to `$src` make it more semantically clear what is happening?

**Comment by aanderse** [comment]
`$src` wasn't patched.

**Comment by Thesola10** [comment]
The original repo mentions a `/etc/crumbs.conf` file, and a quick `strace` confirms that our built `crumbs` accesses it from its derivation path (expected). Should I add configuration entries to generate the file?

**Comment by aanderse** [comment]
As far as I can tell `crumbs.conf` is trivial and provides a very sane default. Given the `crumbs` executable has a way to override the location of `crumbs.conf` I think leaving everything as you currently have it is perfectly reasonable.

**Comment by Thesola10** [comment]
Side note -- why is the "changes requested" blocker still present? Is my branch ready for merging?

**Comment by aanderse** [comment]
@GrahamcOfBorg build crumbs

I guess I was waiting to hear if you were satisfied with my response :smile:

**Review by aanderse — APPROVED** [review]

**Comment by aanderse** [comment]
@Thesola10 thank you for your contribution! :tada:

**Pull request merged**
